Add feasibility builder for explicit constraints

khurram-ghani opened the pull request:
Also includes initial notebook showing penalty-EI method. This notebook will be expanded further in the PR for constrained optimisation.

khurram-ghani committed:
Add support for Box explicit input constraints

Includes two new classes for linear and nonlinear constraints.

khurram-ghani committed:
Add box sampling and tests for feasible points

khurram-ghani committed:
Fix black and mypy issues in new verion

khurram-ghani committed:
Add missed flake8 fix

khurram-ghani committed:
Improve repr() for box search-space

Include constraints and ctol.

khurram-ghani committed:
Improvements from PR review comments

khurram-ghani committed:
Update missing init for ctol types

khurram-ghani committed:
Add residuals method to base search space

khurram-ghani committed:
Simplify parametrize for constraints sampling tests

khurram-ghani committed:
Add feasibility builder for explicit constraints

Also added example notebook showing penalty-EI method.

uri-granta commented on trieste/acquisition/function/function.py:
```diff
@@ -468,6 +468,84 @@ def acquisition(x: TensorType) -> TensorType:
 
+class FastConstraintsFeasibility(SingleModelAcquisitionBuilder[ProbabilisticModel]):
+    """
+    Uses the :func:`fast_constraints_feasibility` function to build a
```
I don't think this needs to mention that it uses `fast_constraints_feasibility` (which after all is an implementation detail)?

```suggestion
    Builder for a probability of feasiblity acquisition function that uses the residuals of 
    direct constraints defined in the search space.
```

khurram-ghani replied:
I copied that style from ProbabilityOfFeasibility, but I can remove it.

khurram-ghani replied:
Done in 1e41a57.

uri-granta commented on trieste/acquisition/function/function.py:
```diff
@@ -468,6 +468,84 @@ def acquisition(x: TensorType) -> TensorType:
+            default is CDF of the Normal distribution with a small scale.
+        """
+        self._search_space = search_space
+        self._smoothing_function = smoothing_function
```
Should we check that `search_space.has_constraints` here and raise an exception if it doesn't?

khurram-ghani replied:
Yes I'll add that in.

khurram-ghani replied:
Done in 1e41a57.

uri-granta commented on trieste/acquisition/function/function.py:
```diff
@@ -468,6 +468,84 @@ def acquisition(x: TensorType) -> TensorType:
+        :param search_space: The global search space over which the feasibility of the constraints
+            is defined.
+        :param smoothing_function: The smoothing function used for constraints residuals. The
+            default is CDF of the Normal distribution with a small scale.
```
(explicitly state the default scale so people don't have to dig through the code?)

khurram-ghani replied:
Done in 1e41a57.

uri-granta commented on tests/unit/acquisition/function/test_function.py:
```diff
@@ -15,7 +15,7 @@
 from __future__ import annotations
```
Do we also want to also add an integration test case to `test_bayesian_optimization.py`? (something similar to the notebook setup perhaps)

khurram-ghani replied:
This will take a bit longer to implement. I'll look into it and maybe add it as part of the next PR where we will have more optimisation methods.

uri-granta commented on docs/notebooks/explicit_constraints.pct.py:
```diff
@@ -0,0 +1,190 @@
+    grid_density=30,
+    contour=True,
+)
+plot_bo_points(query_points, ax[0, 0], num_initial_points, arg_min_idx)
```
The plot is a bit unclear for me: maybe make the dots a bit bigger and the purple one brighter? (Alternatively, ask @vpicheny for a second opinion!)

khurram-ghani replied:
That is an existing plotting function. I don't think the point size can be changed from its args but the colour can be, which I now have. Also, I have made the overall figure larger. Hopefully that is good enough.

uri-granta commented on trieste/acquisition/function/function.py:
```diff
@@ -468,6 +468,84 @@ def acquisition(x: TensorType) -> TensorType:
+    :return: The function for probability of feasibility of constraints.
+    """
+
+    @tf.function
```
Looking at the notebook, it looks like this is being retraced every loop, which is not ideal! I think the issue is because ExpectedConstrainedImprovement calls the constraint function with the query points, which change shape each time. We may be able to specify an input_signature here to handle this. If you can't quickly figure it out don't worry though: I'll raise an issue since I expect it's happening for the normal ProbabilityOfFeasibility too.

khurram-ghani replied:
Yes, `ProbabilityOfFeasibility` has that problem too. I did spend a little bit of time looking at it, but didn't try changing the signature.

uri-granta replied:
I've raised an issue, thanks: https://github.com/secondmind-labs/trieste/issues/661

uri-granta reviewed (commented):
Looks good!

khurram-ghani committed:
Incorporate review comments

Improve documentation text
Raise if search_space has no constraints
Improve plot of BO points in notebook

uri-granta reviewed (approved):
Happy with this

khurram-ghani committed:
Merge branch 'develop' into khurram/explicit_constr_feasibility
